### PR TITLE
ignoreLines feature for filtering the input file

### LINKED ISSUE #133: Support ignoreLines for all plugins: warn, fix, fix and warn
Fix plugin should remove comments when creating a copy; this way users can write expected files without these comments. This is the case when user wants to coimbine warn and fix plugins on same resourceds.

### LINKED ISSUE #221: Support inline fixers for fix and warn plugin
We need to support the following LIT fixers: 
https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/test/clang-tidy/checkers/abseil-time-comparison.cpp

```
// CHECK-MESSAGES: [[@LINE-1]]:7: warning: perform comparison in the time domain [abseil-time-comparison]
// CHECK-FIXES: absl::FromUnixSeconds(x) <= t1;
```

We need to:
1) Support this logic in a fix and warn plugin
2) To do this we need to generate tmp `Expected` test resource 
3) Compare files and check results

### PR DESCRIPTION
**Pull request by sanyavertolet**
feature/ignoreLines

This pull request closes #133 

What's done:
* [ ] Added ignoreLines option for Warn plugin.
* [ ] Added tests for ignoreLines and Warn plugin.
* [ ] Added ignoreLines option for Fix plugin.
* [ ] Added tests for ignoreLines and Fix plugin.
* [ ] Implemented copying a file in Warn plugin in order to filter out lines that match any regex from ignoreLines.

### CODE REVIEW

File: examples/kotlin-diktat/fix/smoke/src/main/kotlin/org/cqfn/save/IgnoreLinesTest/IgnoreLines/save.toml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+[general]
+    tags = ["FixIgnoreLines"]
+    description = "Test for ignoreLines option in Fix"
+    suiteName = "Autofix: Smoke Tests"
+
+[fix]
+    ignoreLines = ["// IGNORE_ME"]
```
Comment (akuleshov7) [inline examples/kotlin-diktat/fix/smoke/src/main/kotlin/org/cqfn/save/IgnoreLinesTest/IgnoreLines/save.toml:7]: may be test with regex here?

Comment (akuleshov7) [reply examples/kotlin-diktat/fix/smoke/src/main/kotlin/org/cqfn/save/IgnoreLinesTest/IgnoreLines/save.toml:7]: and better to test with several different regexes

Comment (sanyavertolet) [reply examples/kotlin-diktat/fix/smoke/src/main/kotlin/org/cqfn/save/IgnoreLinesTest/IgnoreLines/save.toml]: Done

File: save-common/src/commonMain/kotlin/org/cqfn/save/core/plugin/PluginConfig.kt [hunk 2]
```diff
@@ -74,6 +79,7 @@ data class GeneralConfig(
     val runConfigPattern: Regex? = null,
 ) : PluginConfig {
     override val type = TestConfigSections.GENERAL
+    override var ignoreLinesPatterns: MutableList<Regex> = mutableListOf()
```
Comment (akuleshov7) [inline save-common/src/commonMain/kotlin/org/cqfn/save/core/plugin/PluginConfig.kt:82]: why `var` here?

File: save-common/src/commonMain/kotlin/org/cqfn/save/core/plugin/PluginConfig.kt [hunk 3]
```diff
@@ -25,6 +25,11 @@ interface PluginConfig {
      */
     val type: TestConfigSections
 
+    /**
+     * list of regexes to be ignored
+     */
+    var ignoreLinesPatterns: MutableList<Regex>
```
Comment (akuleshov7) [inline save-common/src/commonMain/kotlin/org/cqfn/save/core/plugin/PluginConfig.kt:31]: why `var` here?

File: save-plugins/fix-and-warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fixandwarn/FixAndWarnPluginConfig.kt [hunk 4]
```diff
@@ -26,14 +26,19 @@ data class FixAndWarnPluginConfig(
     override var configLocation: Path = "undefined_toml_location".toPath()
     override val resourceNamePatternStr: String = "(${fix.resourceNamePatternStr})|(${warn.resourceNamePatternStr})"
 
+    @Transient
```
Comment (akuleshov7) [inline save-plugins/fix-and-warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fixandwarn/FixAndWarnPluginConfig.kt:29]: why `@Transient`? I though that we deserialize this value from toml config

File: save-plugins/fix-and-warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fixandwarn/FixAndWarnPluginConfig.kt [hunk 5]
```diff
@@ -49,6 +54,11 @@ data class FixAndWarnPluginConfig(
         return FixAndWarnPluginConfig(
             fix.validateAndSetDefaults(),
             warn.validateAndSetDefaults()
-        ).also { it.configLocation = this.configLocation }
+        ).also {
+            it.configLocation = this.configLocation
+        }
+    }
+    companion object {
+        internal val defaultIgnoreLines: MutableList<Regex> = mutableListOf()
```
Comment (akuleshov7) [inline save-plugins/fix-and-warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fixandwarn/FixAndWarnPluginConfig.kt:62]: Do we really need an extra constant for it? May be you will simply set the default value (`mutableListOf()`) in place, were it is created?

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt [hunk 6]
```diff
@@ -128,20 +130,32 @@ class FixPlugin(
...
-    private fun createTestFile(path: Path, generalConfig: GeneralConfig): Path {
+    private fun createTestFile(
+        path: Path,
+        generalConfig: GeneralConfig,
+        fixPluginConfig: FixPluginConfig): Path {
         val pathCopy: Path = constructPathForCopyOfTestFile(FixPlugin::class.simpleName!!, path)
         createTempDir(pathCopy.parent!!)
 
-        val expectedWarningPattern = generalConfig.expectedWarningsPattern
+        val defaultIgnoreLinesPatterns: MutableList<Regex> = mutableListOf()
```
Comment (akuleshov7) [inline save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt:140]: why is it named as `defaultIgnoreLinesPatterns`? Why "default"? You are adding values to this list

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt [hunk 7]
```diff
@@ -128,20 +130,32 @@ class FixPlugin(
...
         fs.write(fs.createFile(pathCopy)) {
-            fs.readLines(path).forEach {
-                if (expectedWarningPattern == null || !generalConfig.expectedWarningsPattern!!.matches(it)) {
+            fs.readLines(path)
+                .filter { line ->
+                    fixPluginConfig.ignoreLines?.let {
+                        fixPluginConfig.ignoreLinesPatterns.none { pattern -> pattern.matches(line) }
+                    }
+                        ?: run {
+                            defaultIgnoreLinesPatterns.none {regex -> regex.matches(line) }
```
Comment (akuleshov7) [inline save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt:151]: diktat is not working? where are spaces after `{`

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt [hunk 8]
```diff
@@ -128,20 +130,32 @@ class FixPlugin(
...
+        generalConfig.expectedWarningsPattern?.let { defaultIgnoreLinesPatterns.add(it) }
+        generalConfig.runConfigPattern?.let { defaultIgnoreLinesPatterns.add(it) }
 
         fs.write(fs.createFile(pathCopy)) {
-            fs.readLines(path).forEach {
-                if (expectedWarningPattern == null || !generalConfig.expectedWarningsPattern!!.matches(it)) {
+            fs.readLines(path)
+                .filter { line ->
+                    fixPluginConfig.ignoreLines?.let {
+                        fixPluginConfig.ignoreLinesPatterns.none { pattern -> pattern.matches(line) }
```
Comment (akuleshov7) [inline save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt:148]: use `it` in such small lambdas, it looks better with such syntax sugar

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt [hunk 9]
```diff
@@ -128,20 +130,32 @@ class FixPlugin(
...
+        val defaultIgnoreLinesPatterns: MutableList<Regex> = mutableListOf()
+        generalConfig.expectedWarningsPattern?.let { defaultIgnoreLinesPatterns.add(it) }
+        generalConfig.runConfigPattern?.let { defaultIgnoreLinesPatterns.add(it) }
 
         fs.write(fs.createFile(pathCopy)) {
-            fs.readLines(path).forEach {
-                if (expectedWarningPattern == null || !generalConfig.expectedWarningsPattern!!.matches(it)) {
+            fs.readLines(path)
+                .filter { line ->
+                    fixPluginConfig.ignoreLines?.let {
```
Comment (akuleshov7) [inline save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt:147]: what will happen here if `ignoreLines == null`? Sorry, if I am not completely understand this logic

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt [hunk 10]
```diff
@@ -128,20 +130,32 @@ class FixPlugin(
...
         createTempDir(pathCopy.parent!!)
 
-        val expectedWarningPattern = generalConfig.expectedWarningsPattern
+        val defaultIgnoreLinesPatterns: MutableList<Regex> = mutableListOf()
+        generalConfig.expectedWarningsPattern?.let { defaultIgnoreLinesPatterns.add(it) }
+        generalConfig.runConfigPattern?.let { defaultIgnoreLinesPatterns.add(it) }
 
         fs.write(fs.createFile(pathCopy)) {
-            fs.readLines(path).forEach {
-                if (expectedWarningPattern == null || !generalConfig.expectedWarningsPattern!!.matches(it)) {
```
Comment (akuleshov7) [inline save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt:139]: `warningsPattern` will be now included into the `ignoreLinesPatterns`?

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPluginConfig.kt [hunk 11]
```diff
@@ -61,14 +66,26 @@ data class FixPluginConfig(
             this.batchSeparator ?: other.batchSeparator,
             this.resourceNameTestSuffix ?: other.resourceNameTestSuffix,
             this.resourceNameExpectedSuffix ?: other.resourceNameExpectedSuffix,
-        ).also { it.configLocation = this.configLocation }
+            other.ignoreLines?.let {
+                this.ignoreLines?.let { other.ignoreLines.union(this.ignoreLines) } ?: other.ignoreLines
+            }?.toMutableList() ?: this.ignoreLines
```
Comment (akuleshov7) [inline save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPluginConfig.kt:71]: why do you need to convert it to `mutableList`? I thought that you created this field as a mutableList from a very beginning.

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPluginConfig.kt [hunk 12]
```diff
@@ -61,14 +66,26 @@ data class FixPluginConfig(
...
+        ).also {
+            it.configLocation = this.configLocation
+        }
     }
 
+    // due to probable bug in ktoml, ignoreLines = [] and no ignoreLines is ktoml are parsed to be mutableListOf("null")
```
Comment (akuleshov7) [inline save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPluginConfig.kt:77]: https://github.com/akuleshov7/ktoml/issues/77

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPluginConfig.kt [hunk 13]
```diff
@@ -61,14 +66,26 @@ data class FixPluginConfig(
...
         resourceNameTest,
-        resourceNameExpected
-    ).also { it.configLocation = this.configLocation }
+        resourceNameExpected,
+        ignoreLines?.filter { it != "null" }?.toMutableList() ?: ignoreLines
+    ).also {
+        it.configLocation = this.configLocation
+    }
+    companion object {
+        internal val defaultIgnoreLines: MutableList<Regex> = mutableListOf()
```
Comment (akuleshov7) [inline save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPluginConfig.kt:89]: same as above

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt (hunk 7 above)
Comment (akuleshov7) [reply save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt:151]: gradlew.bat diktatFixAll

**Review by akuleshov7 — CHANGES_REQUESTED** [review]
please see comments above

File: save-common/src/commonMain/kotlin/org/cqfn/save/core/plugin/PluginConfig.kt (hunk 2 above)
Comment (sanyavertolet) [reply save-common/src/commonMain/kotlin/org/cqfn/save/core/plugin/PluginConfig.kt:84]: `val` should be here

File: save-common/src/commonMain/kotlin/org/cqfn/save/core/plugin/PluginConfig.kt (hunk 3 above)
Comment (sanyavertolet) [reply save-common/src/commonMain/kotlin/org/cqfn/save/core/plugin/PluginConfig.kt:31]: `val` should be here

File: save-plugins/fix-and-warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fixandwarn/FixAndWarnPluginConfig.kt (hunk 4 above)
Comment (sanyavertolet) [reply save-plugins/fix-and-warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fixandwarn/FixAndWarnPluginConfig.kt:29]: The point it that we deserialize `ignoreLines` from toml config. `ignoreLinesPatterns` is initialized with `ignoreLines` that we parse from config so it is not deserialized directly from toml.

File: save-plugins/fix-and-warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fixandwarn/FixAndWarnPluginConfig.kt (hunk 5 above)
Comment (sanyavertolet) [reply save-plugins/fix-and-warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fixandwarn/FixAndWarnPluginConfig.kt:62]: To my mind it looked nicer. But ok, it seems that we don't need it here.

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt (hunk 7 above)
Comment (sanyavertolet) [reply save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt:163]: Interesting. Seems that this case is ignored by diktat.

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt (hunk 6 above)
Comment (sanyavertolet) [reply save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt:152]: If ignoreLines is not set in config file, I should use `expectedWarningsPattern` and `runConfigPattern` by default. So the values I add to this list are values for default case. In this way I think default is ok here.

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt (hunk 9 above)
Comment (sanyavertolet) [reply save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt:159]: I have two variables: `ignoreLines` and `ignoreLinesPatterns`. We deserialize `ignoreLines` from toml config so it is nullable. Than we use `ignoreLinesPatterns` is a mutable list of Regex and it is not nullable.
In fix plugin there are three cases: `ignoreLines` is empty, `ignoreLines` is null or `ignoreLines` is a list of regular expressions. Here I check if `ignoreLines` is null

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt (hunk 10 above)
Comment (sanyavertolet) [reply save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPlugin.kt:151]: User should explicitly put `warningPattern` into `ignoreLines` in toml config.

File: save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPluginConfig.kt (hunk 11 above)
Comment (sanyavertolet) [reply save-plugins/fix-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fix/FixPluginConfig.kt:71]: The point is that we use .union(), which returns Set.

**Review by akuleshov7 — APPROVED** [review]

**Comment by akuleshov7** [comment]
ok, let's test it

**Pull request merged**